build_projects/shared-build-targets-utils/Publishing/AzurePublisher.cs

```diff
@@ -134,14 +146,62 @@ public IEnumerable<string> ListBlobs(string virtualDirectory)
                 }
             }
 
+            ResetLeaseRenewalTaskState();
+
             throw new Exception($"Unable to acquire lease on {blob}");
         }
 
+        private void ResetLeaseRenewalTaskState()
+        {
+            // Cancel the lease renewal task if it was created
+            if (_leaseRenewalTask != null)
+            {
+                _cancellationTokenSource.Cancel();
+                _leaseRenewalTask = null;
+            }
+        }
+
+        private static void AutoRenewLease(AzurePublisher instance, string blob, string leaseId)
+        {
+            // We will renew the lease every 45 seconds
+            TimeSpan maxWait = TimeSpan.FromSeconds(45);
+            TimeSpan delay = TimeSpan.FromMilliseconds(500);
+            TimeSpan waitFor = maxWait;
+
+            CancellationToken token = instance._cancellationTokenSource.Token;
+            while (true)
+            {
+                // If the task has been requested to be cancelled, then do so.
+                token.ThrowIfCancellationRequested();
+
+                try
+                {
+                    CloudBlockBlob cloudBlob = instance._blobContainer.GetBlockBlobReference(blob);
+                    AccessCondition ac = new AccessCondition() { LeaseId = leaseId };
+                    cloudBlob.RenewLeaseAsync(ac).Wait();
+                    waitFor = maxWait;
+                }
+                catch (Exception e)
+                {
+                    Console.WriteLine($"Retrying lease renewal on {blob}, {e.Message}");
+                    waitFor = delay;
+                }
+
+                // If the task has been requested to be cancelled, then do so.
+                token.ThrowIfCancellationRequested();
+                
+                Thread.Sleep(waitFor);
+            }
+        }
+
         public void ReleaseLeaseOnBlob(string blob, string leaseId)
         {
             CloudBlockBlob cloudBlob = _blobContainer.GetBlockBlobReference(blob);
             AccessCondition ac = new AccessCondition() { LeaseId = leaseId };
             cloudBlob.ReleaseLeaseAsync(ac).Wait();
+
+            // Cancel the lease renewal task
+            ResetLeaseRenewalTaskState();
```

Comment:
My thoughts too, avoid the scenario where that happens as much as possible.

There'll still be the small window of releasing and renewing at the same time if the renewal is already happening at the time of cancellation, although I'm not certain the best way to handle that. Does `.Cancel()` block until the thread actually cancels? Perhaps we should also wait for the thread to join when cancelling if not.